src/server/tran_server.cpp

```diff
@@ -142,45 +140,53 @@ tran_server::boot (const char *db_name)
 void
 tran_server::push_request (tran_to_page_request reqid, std::string &&payload)
 {
-  std::shared_lock<std::shared_mutex> s_lock (m_page_server_conn_vec_mtx);
-  if (m_page_server_conn_vec.empty ())
+  int err_code = NO_ERROR;
+  auto slock = std::shared_lock<std::shared_mutex> { m_main_conn_mtx };
+  while (true)
     {
-      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_CONN_NO_PAGE_SERVER_AVAILABLE, 0);
-      return; // All connections have been disconnected already
+      err_code = m_main_conn->push_request (reqid, std::move (payload));
```

Comment:
> Also, given that m_main_conn might become nullptr in extreme cases (see comment below), verification for null before using it?

Do you mean the case where `reset_main_connection` set it to nullptr, and someone calles `push_request` again without handling the `ER_CONN_NO_PAGE_SERVER_AVAILABLE` error?

Reply:
> Copying the payload can induce a performance hazard but - at least for now - we have to do it.
We will also refactor this to use a more pointer-y approach (rather than moving the underlying data).

Oh, I missed it. Thank you. We may copy the string in this function for now.

> My opinion is: can we withhold processing disconnect requests from a PS as long as a connection handler has in-flight requests? Or does this overcomplicate things?

Actually, I considered a similar approach: Change the disconnection protocol to "All responses are served before handling disconnection request". It may remove the `stop_response_broker` and keep logic simpler. It might be a better approach in terms of safety and simplicity but also has downsides.
- It can be delayed to process the disconnection request.
- There can be a starvation of disconnection if requests are pushed continuously, and it amplifies the delay problem.

So, I've just chosen to keep the existing protocol to make use of existing logic. And about "Copying the payload" problem, it will be still needed to deal with the abnormal disconnection. The `send_recv` can fail abnormally, and it has to be retried somewhere, which requires the `payload` again. 

I lean forward to keep the existing code to make progress. Of course, it has to be fixed if flawed, though. What do you think?